lib/SILOptimizer/UtilityPasses/NonInlinableFunctionSkippingChecker.cpp

```diff
@@ -0,0 +1,107 @@
+//===------------- NonInlinableFunctionSkippingChecker.cpp ----------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2017 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+#include "swift/Basic/LLVM.h"
+#include "swift/SIL/SILFunction.h"
+#include "swift/SIL/SILInstruction.h"
+#include "swift/SIL/SILModule.h"
+#include "swift/SILOptimizer/PassManager/Transforms.h"
+#include "llvm/ADT/StringRef.h"
+#include "llvm/Support/raw_ostream.h"
+
+using namespace swift;
+
+/// Determines whether we should have skipped SILGenning a function that
+/// appears in a SILModule. This only applies when
+/// \c -experimental-skip-non-inlinable-function-bodies is enabled.
+static bool shouldHaveSkippedFunction(const SILFunction &F) {
+  assert(F.getModule().getOptions().SkipNonInlinableFunctionBodies &&
+         "this check only makes sense if we're skipping function bodies");
+
+  // First, we only care about functions that haven't been marked serialized.
+  // If they've been marked serialized, they will end up in the final module
+  // and we needed to SILGen them.
+  if (F.isSerialized())
+    return false;
+
+  // Next, we're looking for functions that shouldn't have a body, but do. If
+  // the function doesn't have a body (i.e. it's an external declaration), we
+  // skipped it successfully.
+  if (F.isExternalDeclaration())
+    return false;
+
+  // Thunks and specializations are automatically synthesized, so it's fine that
+  // they end up in the module.
+  if (F.isThunk() || F.isSpecialization())
+    return false;
+
+  // FIXME: We can probably skip property initializers, too.
+  auto func = F.getLocation().getAsASTNode<AbstractFunctionDecl>();
+  if (!func)
+    return false;
```

Comment:
Yeah, this scares me. All sorts of synthesized things will slip through here too.